src/rpm_ostree/cli_status.rs

```diff
@@ -109,8 +118,31 @@ fn booted_json(status: StatusJSON) -> Fallible<DeploymentJSON> {
     Ok(booted)
 }
 
-/// Introspect deployments (rpm-ostree status).
-fn status_json(booted_only: bool) -> Fallible<StatusJSON> {
+/// Introspect deployments (rpm-ostree status) using rpm-ostree client actor client's
+/// cache if possible.
+fn status_json(client: &mut RpmOstreeClient, booted_only: bool) -> Fallible<StatusJSON> {
+    let ostree_depls_data = fs::metadata(OSTREE_DEPLS_PATH)?;
+    let ostree_depls_data_mtime = FileTime::from_last_modification_time(&ostree_depls_data);
+
+    if let Some(cache) = &client.status_cache {
+        if cache.mtime == ostree_depls_data_mtime {
+            trace!("cache fresh, using cached rpm-ostree status");
+            return Ok(cache.status.clone());
+        }
+    }
+
+    trace!("cache stale, invoking rpm-ostree to retrieve local deployments");
+    let status = invoke_cli_status(booted_only)?;
```

Comment:
I'm mildly concerned by this, as we can technically cache a wrong/incomplete state depending on the boolean that gets here when priming the cache initially.

I'd rather switch to cache the full state and do the booted filtering ourselves, or query `rpm-ostree status` twice in here and cache the booted/non-booted details separately (depending on the ergonomics of the consumers that needs to be massaged).

Reply:
Thanks for pointing that out. I think we should always cache the full state in the `status_json()` function, since I don't think asking `rpm-ostree` to give us the full state would be a lot more expensive than only asking for `--booted`. Currently, I think whenever we use this `status_json()` function, we're always doing `booted_only=false` anyways.

Reply:
SGTM, it looks we can drop the `booted_only` args from the lower layers if that's true, and keep the filtering logic higher up.

While at it (or maybe later as you wish) you may also consider adding metrics regarding total-queries, failed-queries, and some deployments stats. It would probably ease some of the work in https://github.com/coreos/zincati/pull/456 too.

Reply:
Yeah, maybe I'll create a separate PR for the metrics-related things, since I could see a lot of that would depend on the changes in this PR.